### PR TITLE
[DSD] keep 'exp_avg' as DTensor after torch.distributed.checkpoint.state_dict.set_optimizer_state_dict

### LINKED ISSUE #126950: [DSD] keep 'exp_avg' as DTensor after `torch.distributed.checkpoint.state_dict.set_optimizer_state_dict`
### 🚀 The feature, motivation and pitch

command: `torchrun --standalone --nproc_per_node=2 run_fsdp.py`

```
import logging
import os
from typing import Callable, Optional, Tuple
from torch.distributed._tensor import DTensor

import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed.fsdp.wrap import ModuleWrapPolicy
from torch.distributed.checkpoint import state_dict as ptd_state_dict
from torch.distributed._composable.fsdp import fully_shard
import pdb
import sys

def main():
    dist.init_process_group(backend="nccl")
    gpu_id = int(os.environ["LOCAL_RANK"])
    device = f"cuda:{gpu_id}"
    torch.cuda.set_device(device)
    torch.manual_seed(0)
    model = nn.Sequential(*[nn.Linear(4, 4, device=device, bias=False) for _ in range(2)])
    for layer in model:
        fully_shard(layer)
    fully_shard(model)
    optim = torch.optim.Adam(model.parameters(), lr=1e-2)

    x = torch.randn((4, 4), device=device)
    model(x).sum().backward()
    optim.step()
    optim.zero_grad()
    assert isinstance(list(optim.state.values())[0]['exp_avg'], DTensor)
    opt_state_dict = ptd_state_dict.get_optimizer_state_dict(
        model,
        optim,
        options=ptd_state_dict.StateDictOptions(
            full_state_dict=True
        ),
    )
    optim = torch.optim.Adam(model.parameters(), lr=1e-2)
    ptd_state_dict.set_optimizer_state_dict(
        model,
        optim,
        optim_state_dict=opt_state_dict,
        options=ptd_state_dict.StateDictOptions(full_state_dict=True),
    )
    assert isinstance(list(optim.state.values())[0]['exp_avg'], DTensor)


if __name__ == "__main__":
    main()
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mori360**
Fixes #126950
`ptd_state_dict` with `broadcast_from_rank0=False` might miss 2 condition checks in the `set_optimizer_state_dict`
Here we add another condition `full_state_dict=True` with corresponding tensor distribution without broadcasting if broadcast_from_rank0=False

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC @MeetVadakkanchery

### CODE REVIEW

File: torch/distributed/checkpoint/state_dict.py [hunk 1]
```diff
@@ -858,7 +858,7 @@ def _load_optim_state_dict(
                 optim_state_dict = FSDP.optim_state_dict_to_load(
                     model, optim, optim_state_dict
                 )
-        elif info.broadcast_from_rank0:
+        elif (info.broadcast_from_rank0 or info.full_state_dict):
```
Comment (fegin) [inline torch/distributed/checkpoint/state_dict.py:861]: This is not correct. If `info.broadcast_from_rank0` is False, then all ranks have the full_state_dict. We should extract the `full_state_dict` logic from here and make it as a standalone API. And then `info.broadcast_from_rank0` will be come `broadcasting` + `full_state_dict`.

You also need to do a similar thing for model state_dict.

**Review by fegin — CHANGES_REQUESTED** [review]
Please extract the `full_state_dict` logic from `broadcast_from_rank0` and also change the model state_dict.

**Review by fegin — CHANGES_REQUESTED** [review]
Please extract the `full_state_dict` logic from `broadcast_from_rank0` and also change the model state_dict.

File: test/distributed/checkpoint/test_state_dict.py [hunk 2]
```diff
@@ -711,6 +712,38 @@ def test_optim_state_dict_param_matching(self) -> None:
         if dist.get_rank() == 0:
             self.assertTrue("initial_lr" in optim.param_groups[0])
 
+    @with_comms
+    @skip_if_lt_x_gpu(2)
+    def test_optim_state_dict_tensor_matching(self) -> None:
+        device = "cuda"
+        torch.manual_seed(0)
+        model = nn.Sequential(*[nn.Linear(4, 4, device=device, bias=False) for _ in range(2)])
```
Comment (weifengpy) [inline test/distributed/checkpoint/test_state_dict.py:720]: i know it's confusing
`fully_shard` should be `torch.distributed._composable.fsdp.fully_shard`
`torch.distributed._composable.fully_shard` is deprecated

I debug this with `print(model)`

I see
```
Sequential(
  (0): Linear(in_features=4, out_features=4, bias=False)
  (1): Linear(in_features=4, out_features=4, bias=False)
)
```

but it should be if `fully_shard` is applied correctly
```
FSDPSequential(
  (0): FSDPLinear(in_features=4, out_features=4, bias=False)
  (1): FSDPLinear(in_features=4, out_features=4, bias=False)
)
```

Comment (awgu) [reply test/distributed/checkpoint/test_state_dict.py:720]: yes, very sorry about the confusing imports!

File: torch/distributed/checkpoint/state_dict.py (hunk 1 above)
Comment (mori360) [reply torch/distributed/checkpoint/state_dict.py]: Thank you for the comment, now I switch the condition to full_state_dict
And the broadcasting action are conducted under the condition of (broadcast_from_rank0 and rank=0)

**Comment by mori360** [comment]
> Please extract the `full_state_dict` logic from `broadcast_from_rank0` and also change the model state_dict.

`full_state_dict` logic is extracted to distribution functions.

File: torch/distributed/_state_dict_utils.py [hunk 3]
```diff
@@ -585,6 +596,34 @@ def _broadcast_state_dict(
...
+) -> None:
+    # Distribute tensors in each rank
+    ret = {}
+    for key, value in full_state_dict.items():
+        if not torch.is_tensor(value) and key in local_state_dict:
+            local_state_dict[key] = value
+        elif value.dim() == 0 and key in local_state_dict:
+            local_state_dict[key] = value.cpu()
+        else:
+            ret[key] = _TensorInfo(value.size(), value.dtype)
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:613]: Why do you need this? It is not used in any other place.

File: torch/distributed/_state_dict_utils.py [hunk 4]
```diff
@@ -585,6 +596,34 @@ def _broadcast_state_dict(
...
 
 
+def _distribute_state_dict(
+    full_state_dict: Dict[str, Any],
+    local_state_dict: Dict[str, Any],
+    device: torch.device,
+    pg: Optional[dist.ProcessGroup] = None,
+) -> None:
+    # Distribute tensors in each rank
+    ret = {}
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:606]: No need of this variable?

File: torch/distributed/_state_dict_utils.py [hunk 5]
```diff
@@ -585,6 +596,34 @@ def _broadcast_state_dict(
         _broadcast_tensors(ret, local_state_dict, keys, device, pg)
 
 
+def _distribute_state_dict(
+    full_state_dict: Dict[str, Any],
+    local_state_dict: Dict[str, Any],
+    device: torch.device,
+    pg: Optional[dist.ProcessGroup] = None,
+) -> None:
+    # Distribute tensors in each rank
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:605]: Can you add more description for this function?

File: torch/distributed/_state_dict_utils.py [hunk 6]
```diff
@@ -585,6 +596,34 @@ def _broadcast_state_dict(
...
+    for key, value in full_state_dict.items():
+        if not torch.is_tensor(value) and key in local_state_dict:
+            local_state_dict[key] = value
+        elif value.dim() == 0 and key in local_state_dict:
+            local_state_dict[key] = value.cpu()
+        else:
+            ret[key] = _TensorInfo(value.size(), value.dtype)
+            assert isinstance(value, torch.Tensor)
+            full_tensor = value.detach().to(device)
+            local_state = local_state_dict.get(key, None)
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:616]: Do we want to do overwrite? That's what you did in the previous PR for the broadcast from rank0?

File: torch/distributed/_state_dict_utils.py [hunk 7]
```diff
@@ -585,6 +596,34 @@ def _broadcast_state_dict(
...
+    local_state_dict: Dict[str, Any],
+    device: torch.device,
+    pg: Optional[dist.ProcessGroup] = None,
+) -> None:
+    # Distribute tensors in each rank
+    ret = {}
+    for key, value in full_state_dict.items():
+        if not torch.is_tensor(value) and key in local_state_dict:
+            local_state_dict[key] = value
+        elif value.dim() == 0 and key in local_state_dict:
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:610]: Do we want to do `key in local_state_dict` check? Isn't that what you remove in the previous PR to fix the initial_lr issue?

**Review by fegin — COMMENTED** [review]
LGTM, but some changes still need.

File: torch/distributed/_state_dict_utils.py (hunk 3 above)
Comment (mori360) [reply torch/distributed/_state_dict_utils.py:613]: Thank you for the comment, I will remove ret in the next version

File: torch/distributed/_state_dict_utils.py (hunk 5 above)
Comment (mori360) [reply torch/distributed/_state_dict_utils.py:605]: Will add a more detailed description:
Full_state_dict = True but broadcast_from_rank0 = False here. Each rank have full_state_dict. Skip the broadcasting in ``_broadcast_state_dict`` and distribute tensors in each rank

File: torch/distributed/_state_dict_utils.py (hunk 7 above)
Comment (mori360) [reply torch/distributed/_state_dict_utils.py]: After discussion with review, we decide to skip the case if key in local_state_dict

**Review by fegin — APPROVED** [review]
Thanks for fixing the issue!

**Comment by weifengpy** [comment]
nice! congrats to another PR in pytorch. Model init journey ends here and unshard/reshard journey starts

**Pull request closed without merging**